service/history/callback_queue_active_task_executor.go

```diff
@@ -0,0 +1,268 @@
+	"go.temporal.io/server/common/metrics"
+	commonnexus "go.temporal.io/server/common/nexus"
+	"go.temporal.io/server/service/history/callbacks"
+	"go.temporal.io/server/service/history/configs"
+	"go.temporal.io/server/service/history/consts"
+	"go.temporal.io/server/service/history/queues"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/tasks"
+	"go.temporal.io/server/service/history/workflow"
+	wcache "go.temporal.io/server/service/history/workflow/cache"
+)
+
+var retryable4xxErrorTypes = []int{
+	http.StatusRequestTimeout,
+	http.StatusTooManyRequests,
+}
+
+// HTTPCaller is a method that can be used to invoke HTTP requests.
+type HTTPCaller func(*http.Request) (*http.Response, error)
+
+type callbackQueueActiveTaskExecutor struct {
+	taskExecutor
+	config            *configs.Config
+	payloadSerializer commonnexus.PayloadSerializer
+	clusterName       string
+	callerProvider    func(namespaceID, destination string) HTTPCaller
+}
+
+var _ queues.Executor = &callbackQueueActiveTaskExecutor{}
+
+func newCallbackQueueActiveTaskExecutor(
+	shardCtx shard.Context,
+	workflowCache wcache.Cache,
+	logger log.Logger,
+	metricsHandler metrics.Handler,
+	config *configs.Config,
+	callerProvider func(namespaceID, destination string) HTTPCaller,
+) *callbackQueueActiveTaskExecutor {
+	return &callbackQueueActiveTaskExecutor{
+		taskExecutor: taskExecutor{
+			shardContext:   shardCtx,
+			cache:          workflowCache,
+			logger:         logger,
+			metricsHandler: metricsHandler.WithTags(metrics.OperationTag(metrics.OperationCallbackQueueProcessorScope)),
+		},
+		config:         config,
+		clusterName:    shardCtx.GetClusterMetadata().GetCurrentClusterName(),
+		callerProvider: callerProvider,
+	}
+}
+
+func (t *callbackQueueActiveTaskExecutor) Execute(
+	ctx context.Context,
+	executable queues.Executable,
+) queues.ExecuteResponse {
+	task := executable.GetTask()
+	taskType := "Active" + task.GetType().String()
+
+	namespaceTag, replicationState := getNamespaceTagAndReplicationStateByID(
+		t.shardContext.GetNamespaceRegistry(),
+		task.GetNamespaceID(),
+	)
+	metricsTags := []metrics.Tag{
+		namespaceTag,
+		metrics.TaskTypeTag(taskType),
+		metrics.OperationTag(taskType),
+	}
+
+	// We don't want to execute callback tasks when handing over a namespace to avoid starting work that may not be
+	// committed and cause duplicate requests.
+	// We Check namespace handover state **once** when processing is started. Callback tasks may take up to 10
+	// seconds (by default), but we avoid checking again later, before committing the result, to attempt to commit
+	// results of inflight tasks and not lose the progress.
+	if replicationState == enumspb.REPLICATION_STATE_HANDOVER {
+		// TODO: Move this logic to queues.Executable when metrics tags don't need to be returned from task executor.
+		return queues.ExecuteResponse{
+			ExecutionMetricTags: metricsTags,
+			ExecutedAsActive:    true,
```

Comment:
The comment says we don't want to execute the task when handing over, but we still mark `ExecutedAsActive=true` in the response? Seems a little unintuitive to me, but maybe this flag is ignored when there was an execution error?

Reply:
This is copied from another executor.
This whole piece should be refactored.